Seventh app

violentr opened the pull request:
- Save current basket state to the local storage
- Create middleware (which outputs current selected product to console) 
   save current basket state to the local storage
- Create middleware for all action creators. (Basket, Cart)

violentr committed:
Add ability to save and load from localStorage

violentr committed:
Add Save to localStorage as soon as state updated

violentr committed:
Add Basket state read from LocalStorage

violentr committed:
Add logmiddleware when addToBasket btn clicked

Before action dispatched console.log current
product, subscribe to the store and save
basket state to the local storage

violentr committed:
Add Could list individual product both ways

Make sure that individual product accessible
eather way from localstorage or redux store

violentr committed:
Make action declarative in nature, return action

No api request made in action,
Just return one of the action like FETCH_PRODUCTS_REQUEST

violentr committed:
Make action declarative in nature, for individ product

No api request made in action,
Just return one of the action like FETCH_PRODUCT_REQUEST

violentr committed:
Add APIMiddleware, fetch products via api call

Logic for getting list of products from api
done with Promise.

violentr committed:
Add APIMiddleware to the redux store

violentr committed:
Add basketMiddleware, no store subscribe used

index01d commented on Week-1/src/store/index.js:
```diff
@@ -2,7 +2,26 @@ import thunk from 'redux-thunk'
+import APIMiddleware from '~/src/middleware/Api.js'
+import * as BasketActionTypes from '~/src/actiontypes/Basket.js'
 
+const basketMiddleware = (store) => (next) => (action) => {
```
нужно выделить в отдельный модуль, loadState нужно делать здесь

violentr replied:
@index01d не очень понятен комментарий, я понял что функцию надо вынести в отдельный файл
но мне не понятно что надо делать с loadState() ? что я делаю не верно ?

index01d replied:
loadState нужно вызывать в middleware при обработке action'a LOAD_CART, или чего-то подобного. и один раз устанавливать состояние в reducer'e. А затем использовать это состояние в приложении с помощью connect()

index01d replied:
аналогично подобная работа с побочным эффектом (загрузка данных по API или localStorage это побочный эффект) реализована в https://github.com/violentr/reactjs-workshop/blob/seventh_app/Week-1/src/middleware/Api.js

index01d reviewed (changes requested):
Загрузку данных из localStorage нужно делать в middleware, а не в коде компонентов

violentr committed:
Add LOAD_CART actiontype

violentr committed:
Add action creator addToStorage

violentr committed:
Add Basket middleware, save state to localstorage

violentr committed:
Add reducer for LOAD_CART action type

violentr committed:
Add load state from local storage

This happends everytime when page
refreshed saved state will be
retrieved from browser local storage

violentr committed:
Add read from redux store storedItems

Basket items retrieved from local
storage accessible in redux store
under storedItems

violentr committed:
Fix title for app name

index01d commented on Week-1/src/store/index.js:
```diff
@@ -2,26 +2,23 @@ import thunk from 'redux-thunk'
+const middlewares = [thunk, BasketMiddleware, APIMiddleware]
+const store = createStore(reducers, compose(applyMiddleware(...middlewares), DevTools.instrument()))
+
+const persistedState = loadState()
```
можно так, или, что корректнее, нужно прямо в BasketMiddleware перехватывать LOAD_CART и снабжать action списком продуктов из loadState(), после чего с помощью next(action) продолжать обработку. Корректнее, потому что в этом модуле останется только инициализация store'a и ничего более

violentr committed:
Add car middleware, load items from local storage

violentr committed:
Add cart middleware to redux store, triger LOAD_CART

make sure cart middleware added to to the redux store
After the store is created trigger LOAD_CART,
load saved items from local storage

violentr committed:
Update basket actions to reflect the logic

violentr committed:
Update Basket reducer to reflect the logic

violentr committed:
Fix code remove un used code, remove storedItems

No need to use storedItems, items key was reused
instead, no need to clutter the code, less is more

index01d commented on Week-1/src/middleware/cart.js:
```diff
@@ -0,0 +1,8 @@
+import {loadState} from '~/src/shared/cartPersistence.js'
```
коль скоро это используется только в middleware, то можно перенести модуль
```
- middleware
-- cart
--- cartPersistence.js
--- index.js
```

violentr committed:
Move cartPersistance to middleware folder

violentr committed:
Add Read selected product from local storage

Make sure individual product accesible
when page is refreshed is still available.
